src/zip_chunk_provider.rs

```diff
@@ -0,0 +1,239 @@
+// Panics if more than one folder is found
+fn find_region_folder_path<R: Read + Seek>(
+    zip_archive: &mut ZipArchive<R>,
+) -> Result<String, ZipProviderError> {
+    let mut region_prefix = String::from("/");
+    let mut found_region_count = 0;
+    for i in 0..zip_archive.len() {
+        // This unwrap is safe because we are iterating from 0 to len
+        let file = zip_archive.by_index(i).unwrap();
+        let full_path = file.sanitized_name();
+        // file_name() returns None when the path ends with "/.."
+        // we handle that case as an empty string
+        let folder_name = full_path.file_name().unwrap_or_default();
+        if folder_name == "region" {
+            found_region_count += 1;
+            region_prefix = file.name().to_string();
+            // Keep searching after finding the first folder, to make sure
+            // there is only one region/ folder
+        }
+    }
+    if found_region_count == 0 {
+        return Err(ZipProviderError::RegionFolderNotFound);
+    }
+    if found_region_count > 1 {
+        return Err(ZipProviderError::MoreThanOneRegionFolder);
+    }
+
+    Ok(region_prefix)
+}
+
+impl<R: Read + Seek> ZipChunkProvider<R> {
+    pub fn new(reader: R) -> Result<Self, ZipProviderError> {
+        let mut zip_archive = ZipArchive::new(reader)?;
+        let region_prefix = find_region_folder_path(&mut zip_archive)?;
+        let cache = HashMap::new();
+
+        Ok(ZipChunkProvider {
+            zip_archive,
+            region_prefix,
+            cache,
+        })
+    }
+
+    fn region_path(&self, region_x: i32, region_z: i32) -> String {
+        format!("{}r.{}.{}.mca", self.region_prefix, region_x, region_z)
+    }
+
+    pub fn load_chunk(
+        &mut self,
+        chunk_x: i32,
+        chunk_z: i32,
+    ) -> Result<CompoundTag, ChunkLoadError> {
+        let RegionAndOffset {
+            region_x,
+            region_z,
+            region_chunk_x,
+            region_chunk_z,
+        } = RegionAndOffset::from_chunk(chunk_x, chunk_z);
+
+        let mut buf;
+        let buf = if let Some(buf) = self.cache.get_mut(&(region_x, region_z)) {
+            buf
+        } else {
+            let region_path = self.region_path(region_x, region_z);
+
+            let mut region_file = match self.zip_archive.by_name(&region_path) {
+                Ok(x) => x,
+                Err(ZipError::FileNotFound) => {
+                    return Err(ChunkLoadError::RegionNotFound { region_x, region_z })
+                }
+                Err(ZipError::Io(io_error)) => return Err(ChunkLoadError::ReadError { io_error }),
+                Err(e) => panic!("Unhandled zip error: {}", e),
+            };
+
+            let uncompressed_size = region_file.size();
+            buf = Vec::with_capacity(uncompressed_size as usize);
+            region_file.read_to_end(&mut buf)?;
+
+            // Insert into cache
+            self.cache.insert((region_x, region_z), buf.clone());
```

Comment:
Instead of cloning we can pass ownershisp of buf to cache and then call get method.